src/TriangleNodesProvisioner.cpp

```diff
@@ -450,6 +448,69 @@ namespace blitzdg {
         r_y =-xs/Jac;
         s_x =-yr/Jac;
         s_y = xr/Jac;
+
+        // Normals and Face scalings:
+
+        const index_matrix_type& Fmsk = *Fmask;
+        real_matrix_type& Fscal = *Fscale;
+
+        // interpolate geometric factors to face nodes
+        index_type numLocalFaceNodes = NumFaces*NumFacePoints;
+
+        real_matrix_type fxr(numLocalFaceNodes, NumElements), fxs(numLocalFaceNodes, NumElements);
+        real_matrix_type fyr(numLocalFaceNodes, NumElements), fys(numLocalFaceNodes, NumElements);
+
+        for (index_type k=0; k < NumElements; ++k) {
+            index_type count=0;
+            for (index_type f=0; f < NumFaces; ++f) {
+                for (index_type n=0; n < NumFacePoints; ++n) {
+                    fxr(count, k) = xr(Fmsk(n, f), k);
+                    fxs(count, k) = xs(Fmsk(n, f), k);
+                    fyr(count, k) = yr(Fmsk(n, f), k);
+                    fys(count, k) = ys(Fmsk(n, f), k);
+                    ++count;
+                }
+            }
+        }
+
+        // build normals.
+        real_matrix_type& n_x = *nx, n_y = *ny;
+
+        real_matrix_type norm(numLocalFaceNodes, NumElements);
+
+        index_vector_type fid1(NumFacePoints), fid2(NumFacePoints), fid3(NumFacePoints);
+
+        fid1 = ii;
+        fid2 = ii + NumFacePoints;
+        fid3 = ii + 2*NumFacePoints;
+
+        for(index_type k=0; k < NumElements; ++k) {
+            for (index_type i=0; i < NumFacePoints; i++) {
+                // face 1
```

Comment:
It might be worthwhile to think about how to merge these loops with the ones above so you don't need `fxr, fxs, fyr, fys`. I am not sure it that's possible, but it would make for a cleaner implementation. Nothing you need to worry about now though.